src/utils/utils_cache_line.c

```diff
@@ -20,20 +20,26 @@ static void __set_cache_line_invalid(struct ocf_cache *cache, uint8_t start_bit,
 	line_became_invalid = metadata_clear_valid_sec_changed(cache, line,
 			start_bit, end_bit, &line_remains_valid);
 
-	/* If we have waiters, do not remove cache line
-	 * for this cache line which will use one, clear
-	 * only valid bits
-	 */
-	if (!line_remains_valid && !ocf_cache_line_are_waiters(
-			ocf_cache_line_concurrency(cache), line)) {
-		if (line_became_invalid) {
-			env_atomic_dec(&core->runtime_meta->cached_clines);
-			env_atomic_dec(&core->runtime_meta->
-					part_counters[part_id].cached_clines);
-		}
-		ocf_lru_rm_cline(cache, line);
-		ocf_metadata_remove_cache_line(cache, line);
+	if (line_remains_valid) {
+		ENV_BUG_ON(line_became_invalid);
+		return;
```

Comment:
This `ENV_BUG_ON()` is a new addition here, and mixing it into the logic decreases readability.
I'd rather keep it simple and let the compiler do its thing:
```c
ENV_BUG_ON(line_became_invalid && line_became_invalid);

if (line_remains_valid)
	return;
```